cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -257,6 +258,62 @@ MapBuilderBridge::GetLocalTrajectoryData() {
   return local_trajectory_data;
 }
 
+void MapBuilderBridge::HandleTrajectoryQuery(
+    cartographer_ros_msgs::TrajectoryQuery::Request& request,
+    cartographer_ros_msgs::TrajectoryQuery::Response& response) {
+  const auto trajectory_states = GetTrajectoryStates();
+  if (!trajectory_states.count(request.trajectory_id)) {
+    response.status.code = cartographer_ros_msgs::StatusCode::INVALID_ARGUMENT;
+    response.status.message =
+        absl::StrCat("Requested trajectory with ID ", request.trajectory_id,
+                     " doesn't exist.");
+    return;
+  } else if (trajectory_states.at(request.trajectory_id) ==
+             cartographer::mapping::PoseGraphInterface::TrajectoryState::
+                 DELETED) {
+    response.status.code =
+        cartographer_ros_msgs::StatusCode::RESOURCE_EXHAUSTED;
+    response.status.message =
+        absl::StrCat("Requested trajectory with ID ", request.trajectory_id,
+                     " has been deleted.");
+    return;
+  }
+  const cartographer::common::Time min_time = FromRos(request.min_time);
+  const cartographer::common::Time max_time = FromRos(request.max_time);
+  const auto node_poses = map_builder_->pose_graph()->GetTrajectoryNodePoses();
+  for (const auto& node_id_data :
+       node_poses.trajectory(request.trajectory_id)) {
+    if (!node_id_data.data.constant_pose_data.has_value()) {
+      continue;
+    }
+    const cartographer::common::Time node_time =
+        node_id_data.data.constant_pose_data.value().time;
+    if (node_time < min_time || node_time > max_time) {
```

Comment:
self-review: remove min_time/max_time, this logic can be easily done on client side without losing functionality